src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,646 @@
+/*
+ * src/bin/pg_autoctl/supervisor.c
+ *   Supervisor for services run in sub-processes.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+
+#include <inttypes.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "defaults.h"
+#include "fsm.h"
+#include "keeper.h"
+#include "keeper_config.h"
+#include "keeper_pg_init.h"
+#include "log.h"
+#include "monitor.h"
+#include "pgctl.h"
+#include "state.h"
+#include "service.h"
+#include "supervisor.h"
+#include "signals.h"
+#include "string_utils.h"
+
+/*
+ * Supervisor restart strategy.
+ *
+ * The idea is to restart processes that have failed, so that we can stay
+ * available without external intervention. Sometimes though if the
+ * configuration is wrong or the data directory damaged beyond repair or for
+ * some reasons, the service can't be restarted.
+ *
+ * There is no magic or heuristic that can help us decide if a failure is
+ * transient or permanent, so we implement the simple thing: we restart our
+ * dead service up to 5 times in a row, and spend up to 10 seconds or retrying,
+ * and stop as soon as one of those conditions is reached.
+ *
+ * This strategy is inspired by http://erlang.org/doc/man/supervisor.html and
+ * http://erlang.org/doc/design_principles/sup_princ.html#maximum-restart-intensity
+ *
+ *    If more than MaxR number of restarts occur in the last MaxT seconds, the
+ *    supervisor terminates all the child processes and then itself. The
+ *    termination reason for the supervisor itself in that case will be
+ *    shutdown.
```

Comment:
I'm not sure that the code is that different from the comment here. I think our implementation with the stopTime tracking of the service allows to track that we have restarted the service more than MaxR times in the last MaxT seconds... It's not obvious to me that the two ways to explain things are that different?

Reply:
So, the way I read it.
We do: 
- If we restart 5 times without one process staying alive for 15 seconds, we exit.
- If the process has died and we are still trying to restart after 10 seconds, we exit. NOTE: This can happen at the second restart already, if the first restart failed after 11 seconds. We don't count it as successful start in that case, but we have been restarting for more than 10 seconds, so we exit.

Erlang does:
- Restart whenever the process stops. If this has happened more than MaxR in MaxT seconds already, exit instead. Count the number of restarts in the last MaxT seconds. If this is more than MaxR. Exit. NOTE: This logic does not have a notion of "service running time", this is contained in the MaxT. 


That's why I say they are very different. they have the same variables but use them quite differently. 
Personally I like the Erlang logic, as it seems pretty simple and never considers 2 failures as an exit condition.

Reply:
I have used the 15s of running heuristic to implement MaxR/MaxT without having to track all the restart times, that way I don't have to count how many restarts actually happened in the last MaxT interval... but maybe the difference in the approach is then such as to be user visible and harder to explain.

Reply:
Okay, so I have now fixed it with using a ring buffer to keep track of the latest MaxR + 1 restarts. I have left the default values of 5 MaxR in 10 MaxT as the only time this mechanism kicked in for me in development was a startup problem and pg_autoctl would try again and again very quickly, but now only up to 5 times before giving up.

That said we could now say 5 times in  10 mins and this case would be handled properly too. Because we have large timeouts for some operations, such as contacting the monitor when it's not responsive (yet), we might want to change to something like MaxR=5 and MaxT=300 or something in that order of magnitude, what do you think?

Reply:
I think it makes sense to set MaxT to something higher than 10 seconds indeed. I think 300 seconds sounds reasonable.